compiler/rustc_type_ir/src/search_graph/global_cache.rs

```diff
@@ -50,16 +50,21 @@ impl<X: Cx> GlobalCache<X> {
         additional_depth: usize,
         encountered_overflow: bool,
         nested_goals: NestedGoals<X>,
+        concurrency_is_real: bool,
     ) {
         let result = cx.mk_tracked(result, dep_node);
         let entry = self.map.entry(input).or_default();
         if encountered_overflow {
             let with_overflow = WithOverflow { nested_goals, result };
             let prev = entry.with_overflow.insert(additional_depth, with_overflow);
-            assert!(prev.is_none());
+            if !concurrency_is_real {
+                assert!(prev.is_none());
```

Comment:
given that `evaluation_is_concurrent` is a method on the `cx`, call it inside of htis function instead of passing it in.

Also, please check that `cx.get_tracked(prev.result) == result`

Reply:
...I somehow missed that cx is an arg here too.

Reply:
Are we asserting `cx.get_tracked(prev.result) == result` unconditionally or only when non-concurrent or?

Reply:
well, always if `prev` is `Some` :grin: